File: app/src/main/java/se/gustavkarlsson/skylight/android/evaluation/GeomagLocationEvaluator.java
package se.gustavkarlsson.skylight.android.evaluation;

import javax.inject.Inject;

import dagger.Reusable;
import se.gustavkarlsson.skylight.android.models.factors.GeomagLocation;

@Reusable
public class GeomagLocationEvaluator implements ChanceEvaluator<GeomagLocation> {

	@Inject
	GeomagLocationEvaluator() {
	}

	public Chance evaluate(GeomagLocation geomagLocation) {
		Float latitude = geomagLocation.getLatitude();
		if (latitude == null) {
			return Chance.unknown();
		}
		double absoluteLatitude = Math.abs(latitude);
		double chance = (1.0 / 12.0) * absoluteLatitude - (55.0 / 12.0); // 55-67
		if (chance > 1.0) {
			chance = 2.0 - chance;
		}
		return Chance.of(chance);
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/evaluation/GeomagLocationEvaluator.kt
package se.gustavkarlsson.skylight.android.evaluation

import dagger.Reusable
import se.gustavkarlsson.skylight.android.evaluation.Chance.Companion.UNKNOWN
import se.gustavkarlsson.skylight.android.entities.GeomagLocation
import javax.inject.Inject

@Reusable
class GeomagLocationEvaluator
@Inject
constructor() : ChanceEvaluator<GeomagLocation> {

    override fun evaluate(value: GeomagLocation): Chance {
        val latitude = value.latitude ?: return UNKNOWN
        val absoluteLatitude = Math.abs(latitude).toDouble()
        var chance = 1.0 / 12.0 * absoluteLatitude - 55.0 / 12.0 // 55-67
        if (chance > 1.0) {
            chance = 2.0 - chance
        }
        return Chance(chance)
    }
}
